trieste/acquisition/function/multi_objective.py

```diff
@@ -379,3 +384,245 @@ def _update_expected_improvement_fn(
+        self._penalization: Optional[PenalizationFunction] = None
+        self._penalized_acquisition: Optional[AcquisitionFunction] = None
+
+    def prepare_acquisition_function(
+        self,
+        models: Mapping[str, ProbabilisticModelType],
+        datasets: Optional[Mapping[str, Dataset]] = None,
+        pending_points: Optional[TensorType] = None,
+    ) -> AcquisitionFunction:
+        """
+        :param models: The models.
+        :param datasets: The data from the observer. Must be populated.
+        :param pending_points: The points we penalize with respect to.
+        :return: The HIPPO acquisition function.
+        :raise tf.errors.InvalidArgumentError: If the ``dataset`` is empty.
+        """
+        tf.debugging.Assert(datasets is not None, [])
+        datasets = cast(Mapping[str, Dataset], datasets)
+        tf.debugging.Assert(datasets[self._objective_tag] is not None, [])
+        tf.debugging.assert_positive(
+            len(datasets[self._objective_tag]),
+            message=f"{self._objective_tag} dataset must be populated.",
+        )
+
+        acq = self._update_base_acquisition_function(models, datasets)
+        if pending_points is not None and len(pending_points) != 0:
+            acq = self._update_penalization(acq, models[self._objective_tag], pending_points)
+
+        return acq
+
+    def update_acquisition_function(
+        self,
+        function: AcquisitionFunction,
+        models: Mapping[str, ProbabilisticModelType],
+        datasets: Optional[Mapping[str, Dataset]] = None,
+        pending_points: Optional[TensorType] = None,
+        new_optimization_step: bool = True,
+    ) -> AcquisitionFunction:
+        """
+        :param function: The acquisition function to update.
+        :param models: The models.
+        :param datasets: The data from the observer. Must be populated.
+        :param pending_points: Points already chosen to be in the current batch (of shape [M,D]),
+            where M is the number of pending points and D is the search space dimension.
+        :param new_optimization_step: Indicates whether this call to update_acquisition_function
+            is to start of a new optimization step, of to continue collecting batch of points
+            for the current step. Defaults to ``True``.
+        :return: The updated acquisition function.
+        """
+        tf.debugging.Assert(datasets is not None, [])
+        datasets = cast(Mapping[str, Dataset], datasets)
+        tf.debugging.Assert(datasets[self._objective_tag] is not None, [])
+        tf.debugging.assert_positive(
+            len(datasets[self._objective_tag]),
+            message=f"{self._objective_tag} dataset must be populated.",
+        )
+        tf.debugging.Assert(self._base_acquisition_function is not None, [])
+
+        if new_optimization_step:
+            self._update_base_acquisition_function(models, datasets)
+
+        if pending_points is None or len(pending_points) == 0:
```

Comment:
I don't think we shall inherit at all, because "is a" semantics doesn't seem to apply here. For instance, at the moment we consider HIPPO multi-objective only, while LP is single objective only.

I am more inclined to have a common base which both inherit. But as you say typing might make it very difficult to have such a base. THoughts?

Reply:
Actually the main thing that makes it difficult to have a base class is the fact that LocalPenalization is a SingleModelGreedyAcquisitionBuilder while HIPPO is a GreedyAcquisitionFunctionBuilder. I would leave this for now. We can always clean it up later.

Reply:
Yup, didn't touch it yet, and won't, thanks!